File: GOLANG/Composição.go
package main

import "fmt"

type Motor struct {
    Potencia int
}

func (m Motor) DetalhesMotor() string {
    return fmt.Sprintf("Potência do motor: %d HP", m.Potencia)
}

type Carro struct {
    Marca  string
    Modelo string
    Ano    int
    Motor  Motor
}

func (c Carro) Detalhes() string {
    return fmt.Sprintf("Marca: %s, Modelo: %s, Ano: %d, %s", c.Marca, c.Modelo, c.Ano, c.Motor.DetalhesMotor())
}

func main() {
    motor1 := Motor{Potencia: 150}
    
    carro1 := Carro{"Toyota", "Corolla", 2020, motor1}
    carro2 := Carro{"Honda", "Civic", 2018, Motor{Potencia: 140}}
    carro3 := Carro{"Ford", "Mustang", 2022, Motor{Potencia: 450}}

    fmt.Println(carro1.Detalhes())
    fmt.Println(carro2.Detalhes())
    fmt.Println(carro3.Detalhes())
}

File: JAVA/Composição.java
class Motor {
    private int potencia;

    public Motor(int potencia) {
        this.potencia = potencia;
    }

    public int getPotencia() {
        return potencia;
    }

    public String detalhesMotor() {
        return "Potência do motor: " + potencia + " HP";
    }
}

class Carro {
    private String marca;
    private String modelo;
    private int ano;
    private Motor motor;

    public Carro(String marca, String modelo, int ano, Motor motor) {
        this.marca = marca;
        this.modelo = modelo;
        this.ano = ano;
        this.motor = motor;
    }

    public String detalhes() {
        return "Marca: " + marca + ", Modelo: " + modelo + ", Ano: " + ano + ", " + motor.detalhesMotor();
    }

    public static void main(String[] args) {
        Motor motor1 = new Motor(150);

        Carro carro1 = new Carro("Toyota", "Corolla", 2020, motor1);
        Carro carro2 = new Carro("Honda", "Civic", 2018, new Motor(140));
        Carro carro3 = new Carro("Ford", "Mustang", 2022, new Motor(450));

        System.out.println(carro1.detalhes());
        System.out.println(carro2.detalhes());
        System.out.println(carro3.detalhes());
    }
}
